src/backend/distributed/commands/type.c

```diff
@@ -0,0 +1,1366 @@
+	Oid typeOid = InvalidOid;
+	ObjectAddress *address = NULL;
+
+	Assert(stmt->objectType == OBJECT_TYPE);
+
+	typeName = makeTypeNameFromNameList((List *) stmt->object);
+	typeOid = LookupTypeNameOid(NULL, typeName, missing_ok);
+	address = palloc0(sizeof(ObjectAddress));
+	ObjectAddressSet(*address, TypeRelationId, typeOid);
+
+	return address;
+}
+
+
+/*
+ * CompositeTypeStmtToDrop returns, given a CREATE TYPE statement, a corresponding
+ * statement to drop the type that is to be created. The type does not need to exists in
+ * this postgres for this function to succeed.
+ */
+DropStmt *
+CompositeTypeStmtToDrop(CompositeTypeStmt *stmt)
+{
+	List *names = MakeNameListFromRangeVar(stmt->typevar);
+	TypeName *typeName = makeTypeNameFromNameList(names);
+
+	DropStmt *dropStmt = makeNode(DropStmt);
+	dropStmt->removeType = OBJECT_TYPE;
+	dropStmt->objects = list_make1(typeName);
+	return dropStmt;
+}
+
+
+/*
+ * CreateEnumStmtToDrop returns, given a CREATE TYPE ... AS ENUM statement, a
+ * corresponding statement to drop the type that is to be created. The type does not need
+ * to exists in this postgres for this function to succeed.
+ */
+DropStmt *
+CreateEnumStmtToDrop(CreateEnumStmt *stmt)
+{
+	TypeName *typeName = makeTypeNameFromNameList(stmt->typeName);
+
+	DropStmt *dropStmt = makeNode(DropStmt);
+	dropStmt->removeType = OBJECT_TYPE;
+	dropStmt->objects = list_make1(typeName);
+	return dropStmt;
+}
+
+
+/*
+ * CreateTypeDDLCommandsIdempotent returns a list of DDL statements (const char *) to be
+ * executed on a node to recreate the type addressed by the typeAddress.
+ */
+List *
+CreateTypeDDLCommandsIdempotent(const ObjectAddress *typeAddress)
+{
+	List *ddlCommands = NIL;
+	const char *ddlCommand = NULL;
+	Node *stmt = NULL;
+	StringInfoData buf = { 0 };
+	const char *username = NULL;
+
+	Assert(typeAddress->classId == TypeRelationId);
+
+	if (type_is_array(typeAddress->objectId))
+	{
+		/*
+		 * array types cannot be created on their own, but could be a direct dependency of
+		 * a table. In that case they are on the dependency graph and tried to be created.
+		 *
+		 * By returning an empty list we will not send any commands to create this type.
+		 */
+		return NIL;
+	}
+
+	stmt = CreateTypeStmtByObjectAddress(typeAddress);
+
+	/* capture ddl command for recreation and wrap in create if not exists construct */
+	ddlCommand = DeparseTreeNode(stmt);
+	ddlCommand = wrap_in_sql(CREATE_OR_REPLACE_COMMAND, ddlCommand);
+	ddlCommands = lappend(ddlCommands, (void *) ddlCommand);
+
+	/* add owner ship change so the creation command can be run as a different user */
+	username = GetUserNameFromId(get_typowner(typeAddress->objectId), false);
+	initStringInfo(&buf);
+	appendStringInfo(&buf, ALTER_TYPE_OWNER_COMMAND, getObjectIdentity(typeAddress),
+					 quote_identifier(username));
+	ddlCommands = lappend(ddlCommands, buf.data);
+
+	return ddlCommands;
+}
+
+
+/*
+ * wrap_in_sql
+ *
+ * TODO specialize info create_or_replace command
```

Comment:
yeah, this function is only used to wrap a ddl command in a call to `create_or_replace`. Makes more sense to specialize the function instead of having this generic one which is very specific in how to use it.